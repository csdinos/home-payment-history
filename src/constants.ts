export const DB_CONNECTION = 'db_connection'
export const LOGGER = 'logger'
export const REQUEST = 'hph:request'
export const RESPONSE = 'hph:response'
export const NEXT = 'hph:next'