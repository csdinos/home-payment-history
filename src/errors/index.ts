export {HttpError} from './HttpError'
export {BaseError} from './BaseError'