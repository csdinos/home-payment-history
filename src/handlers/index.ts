export {registerPaymentHandlers} from './payment.handler'