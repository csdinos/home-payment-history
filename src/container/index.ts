import {Container} from 'inversify'
import {connectToDatabase} from '@appRoot/utils'
import * as constants from '@appRoot/constants'
import {PaymentController} from '@appRoot/controllers'
import {PaymentService} from '@appRoot/services'

let container: Container = null

// this would normally be another logger, here we use console.log for simplicity
export function setUpContainer(logger: Console): Container {
  logger.log('Creating app container')

  container = new Container({defaultScope: 'Singleton'})

  // contants
  container.bind(constants.DB_CONNECTION).toConstantValue(connectToDatabase())
  container.bind(constants.LOGGER).toConstantValue(console)

  // services
  container.bind(PaymentService).to(PaymentService).inRequestScope()

  // controllers
  container.bind(PaymentController).to(PaymentController).inRequestScope()
  return container
}


export function getContainer(): Container {
  if (!container) {
    throw new Error('Container not initialized')
  }

  return container
}
