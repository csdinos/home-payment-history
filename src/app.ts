import 'reflect-metadata'
import express from 'express'
import {Express} from 'express'
import bodyParser from 'body-parser'
import {setUpContainer} from '@appRoot/container'
import {registerPaymentHandlers} from '@appRoot/handlers'
import {setHeaders} from './middlewares'
import {globalErrorHandler} from '@appRoot/middlewares'

export const startApp = (): Express => {
  const app: Express = express()

  const appContainer = setUpContainer(console)

// app.use((req: EnrichedRequest, res: Response, next: NextFunction) => {
//   req.container = appContainer.createChild()
//   req.container.bind(constants.REQUEST).toConstantValue(req)
//   req.container.bind(constants.RESPONSE).toConstantValue(res)
//   req.container.bind(constants.NEXT).toConstantValue(next)
//   req.on('end', () => {
//     req.container.unbindAll()
//     delete req.container
//     console.log('scoped Container disposed')
//   })
//   next()
// })

  app.use(bodyParser.urlencoded({extended: true}))
  app.use(setHeaders)

  registerPaymentHandlers(app, appContainer)

  app.use(globalErrorHandler)

  return app
}