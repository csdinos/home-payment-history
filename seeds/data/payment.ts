import {Payment} from '../../src/types'

export const paymentSeedData: Array<Payment> = [
  {
    "contractId": 17689,
    "description": "Rent payment",
    "value": 100,
    "time": "2016-12-09T00:00:00.000Z",
    "isImported": false,
    "createdAt": "2016-12-09T12:57:31.393Z",
    "updatedAt": "2016-12-09T12:57:31.393Z",
    "deletedAt": null
  },
  {
    "contractId": 17689,
    "description": "Rent to be paid",
    "value": -100,
    "time": "2016-11-09T00:00:00.000Z",
    "isImported": false,
    "createdAt": "2016-11-09T12:57:09.708Z",
    "updatedAt": "2016-11-09T12:57:09.709Z",
    "deletedAt": null
  },
  {
    "contractId": 17689,
    "description": "Rent to be paid again",
    "value": -150,
    "time": "2016-11-15T00:00:00.000Z",
    "isImported": false,
    "createdAt": "2016-11-09T12:57:09.708Z",
    "updatedAt": "2016-11-09T12:57:09.709Z",
    "deletedAt": null
  },
  {
    "contractId": 17689,
    "description": "Rent paid for this month",
    "value": 120,
    "time": "2016-10-09T00:00:00.000Z",
    "isImported": false,
    "createdAt": "2016-11-09T12:57:09.708Z",
    "updatedAt": "2016-11-09T12:57:09.709Z",
    "deletedAt": null
  },
  {
    "contractId": 17611,
    "description": "Rent to be paid",
    "value": -100,
    "time": "2016-11-09T00:00:00.000Z",
    "isImported": false,
    "createdAt": "2016-11-09T12:57:09.708Z",
    "updatedAt": "2016-11-09T12:57:09.709Z",
    "deletedAt": null
  }
]