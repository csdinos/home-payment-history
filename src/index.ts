process.env.NODE_ENV === 'production' ? require('module-alias/register') : require('dotenv').config()

import {startApp} from '@appRoot/app'
import {destroyActiveConnections} from '@appRoot/utils/dbConnection'

const server = startApp().listen(process.env.EXTERNAL_PORT)
console.log('Let\'s roll')

const releaseResources = async () => {
  await destroyActiveConnections()
  server.close()
}

process.on('SIGINT', releaseResources)
process.on('SIGTERM', releaseResources)