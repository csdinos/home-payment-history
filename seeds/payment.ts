import * as Knex from "knex";
import {paymentSeedData} from './data/payment'

export async function seed(knex: Knex): Promise<void> {
    await knex("payment").del();
    await knex("payment").insert(paymentSeedData);
};
