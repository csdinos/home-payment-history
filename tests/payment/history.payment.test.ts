import {startApp} from '../../src/app'
import * as request from 'supertest'
import {Response} from 'supertest'
import {destroyTestDb, initTestDb} from '../test-utils'
import {StartedTestContainer} from 'testcontainers'
import {paymentSeedData} from '../../seeds/data/payment'

describe('Payment history endpoint scenarios', () => {

  let container: StartedTestContainer

  beforeAll(async () => {
    container = await initTestDb()
  })

  afterAll(async () => {
    await destroyTestDb(container)
  })

  it('When requesting the payment history (all payments), it should retrieve a correct history according the request data', async () => {
    const requestData = {
      contractId: 17689,
      startDate: '2010-12-09T00:00:00.000Z',
      endDate: '2021-01-20T00:00:00.000Z'
    }
    return (await request
        .agent(startApp())
        .get('/payment/history')
        .send(requestData)
        .auth(process.env.BASIC_AUTH_NAME, process.env.BASIC_AUTH_PASS)
        .expect(200)
        .expect((res: Response) => {
          expect(res.body.sum).toEqual(-30)
          expect(res.body.items.length).toEqual(4)
          expect(res.body.items).toMatchSnapshot()
        })
    )
  })

  it('When requesting the payment history with dates matching no payments, it should retrieve an empty list with 0 sum', async () => {
    const requestData = {
      contractId: 17689,
      startDate: '2010-12-09T00:00:00.000Z',
      endDate: '2011-01-20T00:00:00.000Z'
    }
    return (await request
        .agent(startApp())
        .get('/payment/history')
        .auth(process.env.BASIC_AUTH_NAME, process.env.BASIC_AUTH_PASS)
        .send(requestData)
        .expect(200)
        .expect((res: Response) => {
          expect(res.body.sum).toEqual(0)
          expect(res.body.items.length).toEqual(0)
        })
    )
  })

  it('When requesting the payment history with contractId matching no payments, it should retrieve an empty list with 0 sum', async () => {
    const requestData = {
      contractId: 999,
      startDate: '2010-12-09T00:00:00.000Z',
      endDate: '2021-01-20T00:00:00.000Z'
    }
    return (await request
        .agent(startApp())
        .get('/payment/history')
        .auth(process.env.BASIC_AUTH_NAME, process.env.BASIC_AUTH_PASS)
        .send(requestData)
        .expect(200)
        .expect((res: Response) => {
          expect(res.body.sum).toEqual(0)
          expect(res.body.items.length).toEqual(0)
        })
    )
  })

  it('When requesting payment history with dates equal as one record, it should include it in the response', async () => {
    const expectedPayment = paymentSeedData[0]
    const requestData = {
      contractId: expectedPayment.contractId,
      startDate: expectedPayment.time,
      endDate: expectedPayment.time
    }
    return (await request
        .agent(startApp())
        .get('/payment/history')
        .auth(process.env.BASIC_AUTH_NAME, process.env.BASIC_AUTH_PASS)
        .send(requestData)
        .expect(200)
        .expect((res: Response) => {
          expect(res.body.sum).toEqual(expectedPayment.value)
          expect(res.body.items.length).toEqual(1)
          expect(res.body.items).toMatchObject([{
            ...expectedPayment
          }])
        })
    )
  })

  it('After deleting a payment on the range of the request, it should not include it in the history', async () => {
    const requestData = {
      contractId: 17689,
      startDate: '2010-12-09T00:00:00.000Z',
      endDate: '2021-01-20T00:00:00.000Z'
    }
    return (await request
        .agent(startApp())
        .delete('/payment/1')
        .auth(process.env.BASIC_AUTH_NAME, process.env.BASIC_AUTH_PASS)
        .expect(200)
        .expect(async () => {
          return (await request
              .agent(startApp())
              .get('/payment/history')
              .send(requestData)
              .auth(process.env.BASIC_AUTH_NAME, process.env.BASIC_AUTH_PASS)
              .expect(200)
              .expect((res: Response) => {
                expect(res.body.sum).toEqual(-130)
                expect(res.body.items.length).toEqual(3)
                expect(res.body.items).toMatchSnapshot()
              })
          )
        })
    )
  })
})