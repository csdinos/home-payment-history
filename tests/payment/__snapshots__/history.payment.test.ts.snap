// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payment history endpoint scenarios When requesting the payment history (all payments), it should retrieve a correct history according the request data 1`] = `
Array [
  Object {
    "contractId": 17689,
    "createdAt": "2016-12-09T12:57:31.393Z",
    "deletedAt": null,
    "description": "Rent payment",
    "id": 1,
    "isImported": false,
    "time": "2016-12-09T00:00:00.000Z",
    "updatedAt": "2016-12-09T12:57:31.393Z",
    "value": 100,
  },
  Object {
    "contractId": 17689,
    "createdAt": "2016-11-09T12:57:09.708Z",
    "deletedAt": null,
    "description": "Rent to be paid",
    "id": 2,
    "isImported": false,
    "time": "2016-11-09T00:00:00.000Z",
    "updatedAt": "2016-11-09T12:57:09.709Z",
    "value": -100,
  },
  Object {
    "contractId": 17689,
    "createdAt": "2016-11-09T12:57:09.708Z",
    "deletedAt": null,
    "description": "Rent to be paid again",
    "id": 3,
    "isImported": false,
    "time": "2016-11-15T00:00:00.000Z",
    "updatedAt": "2016-11-09T12:57:09.709Z",
    "value": -150,
  },
  Object {
    "contractId": 17689,
    "createdAt": "2016-11-09T12:57:09.708Z",
    "deletedAt": null,
    "description": "Rent paid for this month",
    "id": 4,
    "isImported": false,
    "time": "2016-10-09T00:00:00.000Z",
    "updatedAt": "2016-11-09T12:57:09.709Z",
    "value": 120,
  },
]
`;
